libdrgn/gdbremote.c

```diff
@@ -0,0 +1,491 @@
+// Copyright (c) Daniel Thompson <daniel@redfelineninja.org.uk>
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <ctype.h>
+#include <endian.h>
+#include <inttypes.h>
+#include <netdb.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "gdbremote.h"
+#include "program.h"
+#include "util.h"
+
+#define VERBOSE_PROTOCOL 0
+
+// We don't expect anyone outside this module to need to check for this error
+// by address so, until that changes, we'll keep it static.
+static struct drgn_error drgn_error_end_of_packet = {
+	.code = DRGN_ERROR_OTHER,
+	.message = "cannot read past end of gdbremote packet",
+};
+
+struct gdb_packet {
+	unsigned char buffer[1024];
+	unsigned int buflen;
+};
+
+struct gdb_7bit_iterator {
+	unsigned char *bufp;
+	unsigned int remaining;
+	unsigned char repeat_char;
+	unsigned char run_length;
+};
+
+static char hexchar(uint8_t nibble)
+{
+	assert(nibble < 16);
+
+	if (nibble < 10)
+		return '0' + nibble;
+
+	return 'a' + nibble - 10;
+}
+
+static unsigned char lookup_hexchar(unsigned char c)
+{
+	if (c < 'A')
+		return 0 + (c - '0');
+
+	return 10 + ((c | 0x20) - 'a');
+}
+
+static struct gdb_7bit_iterator gdb_7bit_iterator_init(struct gdb_packet *pkt)
+{
+	struct gdb_7bit_iterator it = {
+		.bufp = &pkt->buffer[1],
+		.remaining = pkt->buflen - 4,
+		.repeat_char = pkt->buffer[0],
+		.run_length = 0,
+	};
+
+	return it;
+}
+
+/*
+ * Extract a single character from the packet currently being processed.
+ *
+ * Handles run length encoding and escapes.
+ *
+ * The packet *must* be checked using gdb_packet_verify_framing() before
+ * processing because we rely on the trailing # to mark the end of the
+ * packet.
+ *
+ * TODO: Provide this with it's own statically allocated error type
+ *       (to clearly indicate end-of-packet)
```

Comment:
From the looks of it, this TODO is done?

Reply:
Yes, will fix.